internal/serialization/schema_service.go

```diff
@@ -19,21 +16,69 @@
 
 package serialization
 
+import (
+	"context"
+	"sync"
+
+	pubserialization "github.com/hazelcast/hazelcast-go-client/serialization"
+)
+
+type SchemaMsg struct {
+	ID         int64
+	ResponseCh chan *Schema
+}
+
 type SchemaService struct {
-	schemaMap map[int64]Schema
+	schemaMap map[int64]*Schema
+	mu        *sync.RWMutex
+	ch        chan<- SchemaMsg
 }
 
-func NewSchemaService() *SchemaService {
+func NewSchemaService(cfg pubserialization.CompactConfig, ch chan<- SchemaMsg) *SchemaService {
 	return &SchemaService{
-		schemaMap: make(map[int64]Schema),
+		schemaMap: MakeSchemasFromConfig(cfg),
+		mu:        &sync.RWMutex{},
+		ch:        ch,
 	}
 }
 
-func (s *SchemaService) Get(schemaId int64) (Schema, bool) {
-	schema, ok := s.schemaMap[schemaId]
+func (s *SchemaService) Schemas() []*Schema {
+	s.mu.RLock()
+	schemas := make([]*Schema, 0, len(s.schemaMap))
+	for _, v := range s.schemaMap {
+		schemas = append(schemas, v)
+	}
+	s.mu.RUnlock()
+	return schemas
+}
+
+func (s *SchemaService) Get(ctx context.Context, schemaId int64) (schema *Schema, ok bool) {
+	// TODO: return error
+	s.mu.RLock()
+	schema, ok = s.schemaMap[schemaId]
+	s.mu.RUnlock()
+	if !ok {
+		rch := make(chan *Schema)
+		s.ch <- SchemaMsg{ID: schemaId, ResponseCh: rch}
+		select {
+		case schema = <-rch:
+			ok = schema != nil
+			if ok {
+				s.putLocal(schema)
+			}
+		case <-ctx.Done():
+			if ctx.Err() != nil {
+				return nil, false
+			}
+		}
+	}
 	return schema, ok
 }
 
-func (s *SchemaService) PutLocal(schema Schema) {
-	s.schemaMap[schema.ID()] = schema
+func (s *SchemaService) putLocal(schema *Schema) {
+	s.mu.Lock()
```

Comment:
In Java I believe there is a check to make sure that if `ok` is True, the schema you have in your local is the same as `schema` as a sanity check. maybe we can add it here as well

Reply:
In the TDD it states that https://hazelcast.atlassian.net/wiki/spaces/IMDG/pages/3976593439/Compact+Schema+Distribution+TDD#Assumptions%2FLimitations
> In case of a collision, we will simply reject the second schema with the same schema id but different content.

I think Go implementation follows the TDD closer, it rejects the new schema in case there's a collision but Java implementation raises an exception.